### PR TITLE
[FilRPCAudit] Report (WIP, Draft)

### PR DESCRIPTION
**Pull request by rvagg**
Framework for final report. @vasco-santos @olizilla feel free to push to this branch if you want to put some basic details in the sections you're looking at. It can be as rough as you like and I'll tidy up later. Or, just file exploration reports in this repo and I can pull from that.

### CODE REVIEW

File: projects/filjs-audit/README.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
+# Lotus / Filecoin RPC and JavaScript Library Audit
+
+**2021-04**
+
+This report captures a condensed summary of a brief audit conducted of the current state of the JavaScript ecosystem surrounding Filecoin and the affordances provided to JavaScript developers via the Lotus RPC API.
+
+## Lotus RPC API
+
+### OpenRPC support
+
```
Comment (olizilla) [inline projects/filjs-audit/README.md:139]:
```suggestion
> JSON based API definition of the lotus json-rpc interfaces that are auto-generated from the code.

The Good: 
- it is genearted from the go rpc implementation, so is always up to date. 
- It could be used to generate api docs and (almost complete) typescript and rust clients.

The Bad: 
- Not yet auto-discoverable or useable with existing tools like https://playground.open-rpc.org (the api definition doc should be discoverable from the api url, but currently isn't, probably due to auth, TBC)

By manually extracting the definition files from the repo you can generate api docs from it:
https://playground.open-rpc.org/?url=https://gist.githubusercontent.com/olizilla/fc6abf836023b6ee662955695f1cfd21/raw/a4cb5aa3c6fbbc47bcf4b92d2370b8b012a810cb/full.json

API client generation is possible for Typescript and rust, but [it's rough](https://github.com/olizilla/lotus-openrpc-client/blob/main/client/typescript/src/index.ts#L280). This could be improved by adding more detail to the lotus openrpc definition (we should do that) but may also require improving the open-roc/generator project

Notably, we cannot define the file import api with open-rpc, as it only specifies how to define JSON payloads, so we will always need a custom step to create a client or docs that includes the file import api.

It could be used to simplify the custom generation code in https://github.com/filecoin-shipyard/js-lotus-client-schema

Exploration: https://github.com/olizilla/lotus-openrpc-client

Last updated: 31 March 2021 _(auto genereated from rpc code changes)_
```

**Comment by olizilla** [comment]
Added a summary for what I know about the open-rpc support so far.

File: projects/filjs-audit/README.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+API client generation is possible for Typescript and rust, but [it's rough](https://github.com/olizilla/lotus-openrpc-client/blob/main/client/typescript/src/index.ts#L280). This could be improved by adding more detail to the lotus openrpc definition (we should do that) but may also require improving the open-roc/generator project
+
+Notably, we cannot define the file import api with open-rpc, as it only specifies how to define JSON payloads, so we will always need a custom step to create a client or docs that includes the file import api.
+
+It could be used to simplify the custom generation code in https://github.com/filecoin-shipyard/js-lotus-client-schema
+
+Exploration: https://github.com/olizilla/lotus-openrpc-client
+
+Last updated: 31 March 2021 _(auto genereated from rpc code changes)_
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md]: being auto generate is great! in [filecoin-shipyard/js-lotus-client-rpc](https://github.com/filecoin-shipyard/js-lotus-client-rpc) we need to run a go tool to update them manually

File: projects/filjs-audit/README.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+This report captures a condensed summary of a brief audit conducted of the current state of the JavaScript ecosystem surrounding Filecoin and the affordances provided to JavaScript developers via the Lotus RPC API.
+
+## Lotus RPC API
+
+### OpenRPC support
+> JSON based API definition of the lotus json-rpc interfaces that are auto-generated from the code.
+
+The Good: 
+- it is genearted from the go rpc implementation, so is always up to date.
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md]:
```suggestion
- it is generated from the go rpc implementation, so is always up to date. 
```

from my experience playing with this auto generated typescript declarations, there is a pain point on optionality. Interfaces are created taking into account that all the parameters are required and if you use this within a typescript project, the compiler will complain that you need to provide all the things, while some of them are optional

File: projects/filjs-audit/README.md [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+### [Filecoin.js](https://github.com/filecoin-shipyard/filecoin.js)
+
+#### Maturity
+
+The current state of the library specified in the README is `ALPHA: things will not work or work incorectly, will break and the API will change! ⚠️`
+
+Taking into account the activity of the repo, it has not been in actively development since November 2020 (Lotus version: ?), relying on an outdated fork of `filecoin-signing-tools-js`. There is a [Roadmap Issue](https://github.com/filecoin-shipyard/filecoin.js/issues/1) where we can see that general Filecoin operations are no yet implemented.
+
+When installing the module via npm latest, I got `"0.0.5-alpha"`even though github master `package.json` shows `"0.0.3-alpha"`.
```
Comment (rvagg) [inline projects/filjs-audit/README.md]: 😱 might be some missing code we need to chase down here, will make a TODO for that

File: projects/filjs-audit/README.md [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+Last updated: 1 Sep 2019
+
+## Ecosystem libraries
+
+### Glif.io
+
+https://www.glif.io/
+
+#### https://github.com/glifio/wallet
+
```
Comment (olizilla) [inline projects/filjs-audit/README.md:355]:
```suggestion
It's a production app. It's not directly re-usable and it's not yet suffciently documented to be an introductory educational resource, but it is useful as a well built and tested example of an app that talks to Lotus, does something useful, and uses a popular UI stack (next.js, react, redux).

- Adding an architecture diagram and an explanation of how the parts fit together would go along way to making wallet an educational resource.
```

File: projects/filjs-audit/README.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+##### filecoin-message-confirmer
+##### filecoin-message
+##### filecoin-number
+
+Was https://github.com/glifio/filecoin-number
+
+##### filecoin-rpc-client
+##### filecoin-wallet-provider
+##### local-managed-provider
+##### react-components
```
Comment (olizilla) [inline projects/filjs-audit/README.md]:
```suggestion
##### [@glif/filecoin-address](https://github.com/glifio/modules/tree/primary/packages/filecoin-address)

> This is a JS implementation of the Filecoin address type, inspired by go-address. It can create new address instances and encode addresses, and it takes care of decoding and validating checksums.

👍 use it. Adopt it into the filecoin-project org to give more visibilty.

##### [@glif/filecoin-message-confirmer](https://github.com/glifio/modules/tree/primary/packages/filecoin-message-confirmer)

Poll the StateSearchMsg API method to verify that a specfied message CID appears on-chain with a 0 exit-code reciept. Uses `@glif/filecoin-rpc-client`.

A reasonable solution given the current API. We should explore a better API for waiting for a message to succesfully execute.

Need to fix https://github.com/glifio/modules/issues/65

##### [@glif/filecoin-message](https://github.com/glifio/modules/tree/primary/packages/filecoin-message)

Normalises the casing of message object keys between lotus and zondax. If you need to talk to both, you are going to want to avoid dealing with the fact that lotus uses upper-camel case `GasFeeCap` and zondax uses all lower case `gasfeecap`.

##### [@glif/filecoin-number](https://github.com/glifio/modules/tree/primary/packages/filecoin-number)

A wrapper class for bignumber.js to represent an amount of FIL which may be larger than `Number.MAX_SAFE_INTEGER` (2^53 - 1). Provides conversion from attoFIL to picoFIL to FIL, and a Currency Conversion api which uses a remote service to provide the exchange rate.

Review and potentially replace with a lighter / more ergonimic alternative. 

- We should explore the ideal API for this representing FIL amounts.
- Consider using a lighter lib for big number support https://github.com/MikeMcl/big.js/wiki
- Consider just using the built-in BigInt https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
- Currency conversion should be a seperate utility as many use-cases will not need it.

##### [@glif/filecoin-rpc-client](https://github.com/glifio/modules/tree/primary/packages/filecoin-rpc-client)

**TL;DR** Too bare-bones for general consumption, but a good example for folks who like it minimal.

A very light [(~70 LoC) wrapper](https://github.com/glifio/modules/blob/9b97f31054c20b39b8a75d30727fa33162715dab/packages/filecoin-rpc-client/src/index.ts
) around `axios` to create JSON RPC requests.

- 👍 lightweight, low maintenance, never out of date, easy to understand.
- 👍 lovely docs https://documenter.getpostman.com/view/4872192/SWLh5mUd?version=latest
- ❌ methods passed as string arg. No api, no types, no way to discover the api from your IDE.
- ❌ you've got to parse filecoin responses manually
- ❌ CID params are awkward. Must be passed as a ipld-json link `{ '/': '<cid string>' }`


##### [@glif/filecoin-wallet-provider](https://github.com/glifio/modules/tree/primary/packages/filecoin-wallet-provider)

Abstraction layer for filecoin wallet implementations. A good example to explore, but needs more iterations before it's something we should recommend to folks, and it's unlikely to get that attention in the near future.

Aims at being a facade over different filecoin impl providers, to provide a common api and re-use common / computable methods where possible. Ideally a povider impl adds implementations of sensitive functions / primatives while non-senstive / high-level ones are provided by this module.

##### [@glif/react-components](https://github.com/glifio/modules/tree/primary/packages/react-components)

Not for general consumption. Has some nice UI components, but folks can explore the wallet app to see them, and get the most up to date versions.
```

File: projects/filjs-audit/README.md [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+
+### Glif.io
+
+https://www.glif.io/
+
+#### https://github.com/glifio/wallet
+
+#### https://github.com/glifio/modules/tree/primary/packages
+
+Dependents of https://github.com/glifio/wallet, some are also useful by themselves.
```
Comment (olizilla) [inline projects/filjs-audit/README.md]:
```suggestion
### Glif.io

https://www.glif.io/

_interoperable tools for the Filecoin network_

> Exploration report: https://github.com/protocol/w3dt-sudo/pull/19
> Last updated 19 April 2021 

#### https://github.com/glifio/wallet

> A lightweight web interface to send and receive Filecoin via your Ledger device

It's a production app. It's not directly re-usable and it's not yet suffciently documented to be an introductory educational resource, but it is useful as a well built and tested example of an app that talks to Lotus, does something useful, and uses a popular UI stack (next.js, react, redux).

- Adding an architecture diagram and an explanation of how the parts fit together would go along way to making wallet an educational resource.

#### https://github.com/glifio/modules/tree/primary/packages

Dependents of https://github.com/glifio/wallet, some are also useful by themselves. It's a lerna repo, commits every month, and the project builds and works out of the box. Each module is explored below

```

**Comment by rvagg** [comment]
Rendered @ https://github.com/protocol/w3dt-sudo/blob/rvagg/project-filecoin-js/projects/filjs-audit/README.md

File: projects/filjs-audit/README.md [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+* ***(NEEDS INVESTIGATION)*** Look into what is happening with the versioning and whether there's code not pushed to GitHub.
+
+_TODO: are these recommendations?_
+
+- Exported `WsJsonRpcConnector` and `HttpJsonRpcConnector` seem the same as [@filecoin-shipyard/lotus-client-provider-nodejs](@filecoin-shipyard/lotus-client-provider-nodejs) + [@filecoin-shipyard/lotus-client-provider-browser](@filecoin-shipyard/lotus-client-provider-browser), which abstracts the WS or HTTP, but also supports both.
+  - It is worth pointing out that even in the browser, `node-fetch` seems to be bundled and every time the window existence is checked.
+- Interact with traditional wallet providers like metamask (could not get it up and running)
+  - Needs a frictionless experience or a general improvement on the documentation for this core functionality provided
+
+#### Experiments
```
Comment (rvagg) [inline projects/filjs-audit/README.md]: This section seems like the weak point of the report at the moment because it's so much detail compared to the rest. Maybe we could trim this down a bit and link back to the original exploration report? I don't think we want to remove all of it because there's interesting stuff in here.

Comment (rvagg) [reply projects/filjs-audit/README.md]: (actually the real "weak point" is all the unfilled sections, the things we didn't even look at! but in terms of the structure and content of the meaty bits of the report this stands out as different due to its density).

Comment (vasco-santos) [reply projects/filjs-audit/README.md]: yeah, I will link this to the individual exploration report.

> actually the real "weak point" is all the unfilled sections

Yeah, I could not find a better structure for those

File: projects/filjs-audit/README.md [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+
+- Supports Wallet Mnemonics
+- The roadmap indicates that more general Filecoin operations might get implemented in the future.
+- Tests are failing to run, even by trying to run a shell script as done in CI. In the latest commits, CI did also not run.
+
+#### Recommendations
+
+* ***(NEEDS INVESTIGATION)*** Look into what is happening with the versioning and whether there's code not pushed to GitHub.
+
+_TODO: are these recommendations?_
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md]:
```suggestion
```

@rvagg  Rephrased them to better picture recommendations.

File: projects/filjs-audit/README.md (hunk 8 above)
Comment (rvagg) [reply projects/filjs-audit/README.md]: 👌 I certainly don't want to diminish the work and the content is great but the risk is that we overload this with content and nobody reads any of it at all

File: projects/filjs-audit/README.md [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+6. `transactionSerializeRaw`
+7. `transactionParse`
+8. `transactionSign`
+9. `transactionSignLotus`
+10. `transactionSignRaw`
+11. `verifySignature`
+12. `addressAsBytes`
+13.  `bytesToAddress`
+
+All available via Rust, WASM (browser and Node.js) and JSON RPC Server.
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md]:
```suggestion
It includes libraries in `Rust Native Library`, `WASM Library` and `JSON RPC` with the main cryptographic primitives to use in filecoin. More details in the [Exploration Report](https://github.com/protocol/w3dt-sudo/blob/master/home/vasco-santos/2021-04-exploration-report-fil-signing-tools.md)
```

Comment (vasco-santos) [reply projects/filjs-audit/README.md]: This should be changed once #23 is merged in

File: projects/filjs-audit/README.md [hunk 11]
```diff
@@ -0,0 +1,564 @@
...
+
+
+## High-level Summary of Findings
+
+**[OpenRPC support](https://github.com/filecoin-project/lotus/tree/master/build/openrpc)** support was grant-funded and landed in Lotus. It should allow clearer and auto-generated documentation of the API and auto-generation of clients in a variety of languages. Current status: it's basic and needs more work to be useful.
+  * Code documentation extracted for API docs could be greatly improved.
+  * Optional parameters are not identified, causing problems for typed language clients.
+  * Lotus auth currently prevents auto doc and code generation from a live node.
+  * The file import API is separate and difficult to integrate, so we have two APIs.
+  * We should then js-lotus-client into an OpenRPC consumer to dogfood this and simplify our codegen paths.
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md]:
```suggestion
  * We should turn the js-lotus-client into an OpenRPC consumer to dogfood this and simplify our codegen paths.
```

File: projects/filjs-audit/README.md [hunk 12]
```diff
@@ -0,0 +1,564 @@
...
+
+**[js-lotus-client](https://github.com/filecoin-shipyard/js-lotus-client/)** is the most mature JavaScript API client for Lotus and can be used in Node.js and the browser. It uses a custom codegen solution to extract the API, docs and TypeScript definitions.
+  * We need to invest in keep this updated, perhaps automatic releases and switch to OpenRPC codegen.
+  * File import API poses usability challenges and is unclear to users.
+  * Building out a suite of clear and practical example applications would be an excellent investment.
```
Comment (vasco-santos) [inline projects/filjs-audit/README.md:69]:
```suggestion
  * Building out a suite of clear and practical example applications would be an excellent investment.
  * Optional parameters are not identified, causing problems for typed language clients.
```

Comment (vasco-santos) [reply projects/filjs-audit/README.md:69]: same problem as above, as code is generated from the go code

**Review by vasco-santos — CHANGES_REQUESTED** [review]
